### PR TITLE
`ultralytics 8.0.233` improve Classify train augmentations

### LINKED ISSUE #6917: Classification augmentations
## Summary
Execute research and testing on new/additional augmentations to use for Classification task.

## Subtasks
- [x] research new/additional augmentations for image classification
- [x] identified new augmentations to test with
- [x] execute new augmentations
- [x] collect and analyze new augmentation performance impacts
- [x] decide which augmentations to include
- [x] implement new augmentations into deployment code
- [ ] release new iteration of classification task with additional/new augmentations

### PR DESCRIPTION
**Pull request by fcakyon**
### Summary
🧪🛠️🌈

This pull request improves the classification augmentations and transforms for the `ultralytics` package. It adds support for different torchvision versions and new configuration options for auto augmentation, random erasing and color jitter. It also updates the existing classes and functions that use the classification transforms and adds a new test function. The files affected are `ultralytics/data/augment.py`, `ultralytics/data/dataset.py`, `ultralytics/engine/predictor.py`, `ultralytics/models/yolo/classify/predict.py`, `ultralytics/utils/torch_utils.py`, `tests/test_python.py` and `ultralytics/cfg/default.yaml`.

> _`torchvision` augments_
> _classification transforms_
> _better in the fall_

### Walkthrough
*  Add and modify classification transforms for training and evaluation ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L17-R25), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L798-R807), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528R816-R978), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L815-R995), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534L39-R39), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534L210-R210), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9bd07d929cb8522458c879bd81a82d9af468ffbc215e762f39b2a50a601758d7R18), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9bd07d929cb8522458c879bd81a82d9af468ffbc215e762f39b2a50a601758d7R30-R32))
  - Import `torchvision` and define constants for checking torchvision version in `torch_utils.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9bd07d929cb8522458c879bd81a82d9af468ffbc215e762f39b2a50a601758d7R18), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9bd07d929cb8522458c879bd81a82d9af468ffbc215e762f39b2a50a601758d7R30-R32))
  - Rename `classify_transforms` to `classify_transforms_train` and add more parameters and functionality in `augment.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L798-R807), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L815-R995))
  - Add `classify_transforms_eval` to create evaluation transforms in `augment.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528R816-R978))
  - Import and use the new transform functions in `predictor.py` and `dataset.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534L39-R39), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534L210-R210), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L11-R17), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L226-R246))
* Add and modify configuration options for classification augmentations in `default.yaml` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-732d61cb4969eff986cec58058000b2949ea3b429a97ca5a1045897e754145c5R110-R112))
  - Pass the arguments from the `args` object to the transform functions in `dataset.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L226-R246))
* Add and modify tests for classification transforms in `test_python.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-51deb1ccb3a8b618784df2e2abfbc3b056436f1db72d1896d5102f9906978492R446-R487))
* Modify the `ClassificationDataset` and `ClassificationPredictor` classes to handle the new and legacy transforms ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L236-R253), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L251-R273), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL3-R5), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL31-R44))
  - Use `DEFAULT_MEAN` and `DEFAULT_STD` instead of hard-coded values in `dataset.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L236-R253))
  - Convert the NumPy array to a PIL image before applying the torch transforms in `dataset.py` and `predict.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L251-R273), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL3-R5), [link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL31-R44))
  - Check and store the legacy transform name in `predict.py` ([link](https://github.com/ultralytics/ultralytics/pull/4546/files?diff=unified&w=0#diff-663b5aaaf7a04b1a7fe40326af039d2baf50df166e9738332eacbae9271ec55fL31-R44))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhanced classification model performance with improved augmentation settings.

### 📊 Key Changes
- 🔧 Updated accuracies for YOLOv8 classification models in documentation files.
- 🎨 Added new image augmentations with random erasing and auto augmentation policies like `randaugment`, `augmix`, and `autoaugment`.
- ✅ Improved and added new tests for classification augmentation transformations.
- 📈 Updated the default configuration to include new augmentation settings.

### 🎯 Purpose & Impact
- 🎯 The purpose is to improve model accuracies and provide more robust training through advanced augmentation techniques.
- 💥 This could lead to better-performing models and more variability in training, potentially improving the model's generalization capabilities. Users can now experiment with different augmentation settings tailored to their specific needs for more effective training customizations.

### CODE REVIEW

**Comment by fcakyon** [comment]
I have read the CLA Document and I sign the CLA

File: ultralytics/data/augment.py [hunk 1]
```diff
@@ -803,23 +806,116 @@ def classify_transforms(size=224, mean=(0.0, 0.0, 0.0), std=(1.0, 1.0, 1.0)):  #
...
+    if auto_augment:
+        assert isinstance(auto_augment, str)
+        # color jitter is typically disabled if AA/RA on,
+        # this allows override without breaking old hparm cfgs
+        disable_color_jitter = not force_color_jitter
+        if isinstance(size, (tuple, list)):
+            img_size_min = min(size)
+        else:
+            img_size_min = size
+        aa_params = dict(
```
Comment (fcakyon) [inline ultralytics/data/augment.py:867]: TODO: get rid of this

File: ultralytics/data/augment.py [hunk 2]
```diff
@@ -803,23 +806,116 @@ def classify_transforms(size=224, mean=(0.0, 0.0, 0.0), std=(1.0, 1.0, 1.0)):  #
...
+            translate_const=int(img_size_min * 0.45),
+            img_mean=tuple([min(255, round(255 * x)) for x in mean]),
+        )
+
+        if auto_augment == 'randaugment':
+            secondary_tfl += [T.RandAugment(auto_augment, aa_params)]
```
Comment (fcakyon) [inline ultralytics/data/augment.py:873]: TODO: fix inputs

File: ultralytics/data/augment.py [hunk 3]
```diff
@@ -803,23 +806,116 @@ def classify_transforms(size=224, mean=(0.0, 0.0, 0.0), std=(1.0, 1.0, 1.0)):  #
...
+        elif auto_augment == 'augmix':
+            aa_params['translate_pct'] = 0.3
+            secondary_tfl += [T.AugMix(auto_augment, aa_params)]
+        elif auto_augment == 'autoaugment':
+            secondary_tfl += [T.AutoAugment(auto_augment, aa_params)]
```
Comment (fcakyon) [inline ultralytics/data/augment.py:878]: TODO: fix inputs

**Comment by glenn-jocher** [comment]
@fcakyon yes we don't want to apply normalization to any of our YOLO models including classify. Please remove this and fix all CI issues.

File: ultralytics/data/dataset.py [hunk 4]
```diff
@@ -225,7 +227,22 @@ def __init__(self, root, args, augment=False, cache=False, prefix=''):
         self.cache_disk = cache == 'disk'
         self.samples = self.verify_images()  # filter out bad images
         self.samples = [list(x) + [Path(x[0]).with_suffix('.npy'), None] for x in self.samples]  # file, index, npy, im
-        self.torch_transforms = classify_transforms(args.imgsz)
+        scale = (1.0 - args.scale, 1.0),  # (0.08, 1.0)
```
Comment (fcakyon) [inline ultralytics/data/dataset.py:230]:
```suggestion
        scale = (1.0 - args.scale, 1.0)  # (0.08, 1.0)
```

**Comment by fcakyon** [comment]
@glenn-jocher tests are printing 6000 lines of tensor values, im having a hard time finding the line raising error inside test logs :sweat_smile: do you have any suggestions?

**Comment by glenn-jocher** [comment]
@fcakyon the repo is 100% backwards compatible to torch 1.8.0 as in the tests in https://github.com/ultralytics/ultralytics/actions/runs/5984441890/job/16235611100?pr=4546 so we can only accept updates that allow for 1.8.0 to be used.

But that doesn't seem to be the problem because the torch latest tests are also failing in https://github.com/ultralytics/ultralytics/actions/runs/5984441890/job/16235611032?pr=4546

The easiest way to check is to just isolate a single test that's failing and run it directly, i.e. test_python.py test_results() function is failing, so you can debug this function individually like this in PyCharm:


Screenshot 2023-08-26 at 13 10 55

**Comment by glenn-jocher** [comment]
@fcakyon also for checking installed torch versions the method to create a global boolean here and then use it whereever it's needed. You can do the same for the matching torch or torchvision versions you need for the new augmentations and then only apply them, i.e. `if TORCHVISION_0_11`

https://github.com/ultralytics/ultralytics/blob/e9f596430f448e14b13ad1727e42731f7654b8db/ultralytics/utils/torch_utils.py#L27-L29

**Comment by fcakyon** [comment]
I know the error line for tv>0.9. I have to update the self.model.model.transforms during classify prediction. But not able to find the related line where we set the transforms property of the ClassificationModel.

Do you have any idea @glenn-jocher ?

**Comment by fcakyon** [comment]
@glenn-jocher its ready

**Comment by glenn-jocher** [comment]
@fcakyon tests are passing! :)

**Comment by fcakyon** [comment]
@glenn-jocher btw this pr not only improves the training augmemtations but also implements much better eval augmentations!

**Comment by glenn-jocher** [comment]
@fcakyon testing in https://wandb.ai/glenn-jocher/Classify-aug-test

**Comment by glenn-jocher** [comment]
@fcakyon I was reviewing this and I see that all of the argument changes are missing from the docs. The docs explain the arguments for the different modes in a few places, i.e.:

https://docs.ultralytics.com/modes/train/#arguments
https://docs.ultralytics.com/usage/cfg/

Can you please update the pages in the docs/ directory?

**Comment by glenn-jocher** [comment]
@fcakyon hey there's something I don't understand here. Why do we have a new transforms function, `classify_transforms_eval`? 

Is it required, i.e. why don't we continue to use the existing `classify_transforms()` function instead of creating a second function?

**Comment by glenn-jocher** [comment]
@fcakyon we need to rename `crop_percentage` to `crop_fraction` here. A percent ranges from to 0-100, but this is a fraction since it ranges from 0-1. It should also be added the FRACTIONS list here for error checking it's values:

https://github.com/ultralytics/ultralytics/blob/465df3024f44fa97d4fad9986530d5a13cdabdca/ultralytics/cfg/__init__.py#L67-L69

**Comment by glenn-jocher** [comment]
@fcakyon the main problem with this PR is that it's introducing duplicate methods/classes for existing functionality. This is a general tendency of most user PRs unfortunately, but more code and more functions is not better, we want a minimal change and we want to update and utilize existing functions and classes wherever possible. 

In ultralytics/data/augment.py we want only 2 transforms classes: one for val/predict that does not do augmentation, and then a second that applies augmentation for training. We also need the same names for the classes due to backwards compatibility reasons, so we want to update the existing transforms rather than create new transforms. But this PR is adding 3 new transform classes, so imagine trying to explain to a user why we have 5 types of classification transforms and what the differences between them all are.

The best way to think about this is that if eventually we have 100 PRs from 100 users, we don't want 100 new files and functions created, we ideally want these PRs to reduce code (best possible scenario), but if not then to update existing functions and classes, and only in the worst case scenario to introduce additional files and functions, since these expand our maintenance requirements going forward.

Let's take a closer look here and see if we can update the existing transforms with the new functionality.

File: ultralytics/data/augment.py [hunk 5]
```diff
@@ -957,23 +962,184 @@ def classify_transforms(size=224, rect=False, mean=(0.0, 0.0, 0.0), std=(1.0, 1.
...
+):
+    """
+    Classification transforms for evaluation/inference. Inspired by timm/data/transforms_factory.py.
+
+    Args:
+        size (int): image size
+        mean (tuple): mean values of RGB channels
+        std (tuple): std values of RGB channels
+        interpolation (T.InterpolationMode): interpolation mode. default is T.InterpolationMode.BILINEAR.
+        crop_percentage (float): percentage of image to crop. default is 0.875.
```
Comment (glenn-jocher) [inline ultralytics/data/augment.py:981]: @fcakyon it says here that default is 0.875 but in defaults.yaml we have 1.0

**Comment by glenn-jocher** [comment]
@Laughing-q @fcakyon hey guys, I've merged main into this branch and reviewed everything. It seems ready to merge, but we'd need to coordinate this with updating the -cls models in the assets repo to the newly trained models. I was thinking older version of `ultralytics` that download the latest -cls release assets after updating will cause crashes right?

**Comment by Laughing-q** [comment]
@glenn-jocher just tested a new cls model using current main branch and it works properly without any error or crash, but with a slightly top1-acc drop, probably because of the different preprocessing.

@fcakyon I didn't really update the augmentation parts except cleaned some functions that not used, probably we don't need to train new cls models again? please confirm. :)

**Comment by fcakyon** [comment]
@Laughing-q based on my earlier results, top1 accuracies was better than the current cos scores. Maybe the updates in the preprocess affected it. I need to check the new implementation again.

**Comment by Laughing-q** [comment]
@fcakyon yes please confirm. :) 
I just tested the new version of yolov8n-cls model and got the same top1-acc 68.9 vs 68.94 in the wandb dashborad.
屏幕截图 2023-12-25 215759

**Comment by glenn-jocher** [comment]
@fcakyon was reviewing the code here, `crop_fraction` shows up as a predict and val argument, but I think this may be incorrect.

**No augmentation arguments should be set during predict and val**, instead classification models are saved with their `transforms` that are then applied during predict/val. 

Any manual change to any augmentation hyperparameter like `crop_fraction` during predict/val that deviates it from the value used during training will result in worse performance, so there's no reason we should incentivize users to start changing these in predict/val mode. Does this sound right or am I missing something?

**Comment by fcakyon** [comment]
@glenn-jocher @Laughing-q i believe this PR is ready to merge 👍

**Comment by glenn-jocher** [comment]
@fcakyon ok looks good! PR merged for deployment in new 8.0.233!
